### PR TITLE
Fix zoom issue with too thin lines

### PR DESCRIPTION
**Pull request by calixteman**
- aims to fix issue #12868: apply zoom factor to linewidth after setting it to 1.
 - only apply 1px-width when required
 - the sign of getSinglePixelWidth is used to know if 1px-width is required

### CODE REVIEW

File: src/display/canvas.js [hunk 1]
```diff
@@ -922,6 +922,11 @@ const CanvasGraphics = (function CanvasGraphicsClosure() {
       }
       this.ctx.transform.apply(this.ctx, viewport.transform);
 
+      this.viewportTransform = viewport.transform;
+      // Square of viewport scale factor.
+      this.viewportSqScaleFactor =
+        viewport.transform[0] ** 2 + viewport.transform[2] ** 2;
+      this.viewportScaleFactor = Math.sqrt(this.viewportSqScaleFactor);
```
Comment (Snuffleupagus) [inline src/display/canvas.js:929]: This doesn't actually take the optional `transform` argument into account, so I'd expect that you actually want to use `this.baseTransform` in your calculations here?

Furthermore, you're not actually using most of these properties anywhere but here, hence please use local variables where appropriate!

It should be possible to simplify all of this code, after moving it down a couple of lines, to e.g. the following:
```suggestion
      this._combinedScaleFactor =
        Math.sqrt(this.baseTransform[0] ** 2 + this.baseTransform[2] ** 2);
```

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: src/display/canvas.js (hunk 1 above)
Comment (calixteman) [reply src/display/canvas.js]: I forgot to remove them.
Anyway, `baseTransform` is likely better than `viewport.transform`: thanks for the tip.

File: src/display/canvas.js [hunk 2]
```diff
@@ -1357,22 +1361,24 @@ const CanvasGraphics = (function CanvasGraphicsClosure() {
...
           ctx.restore();
         } else {
           const lineWidth = this.getSinglePixelWidth();
-          if (lineWidth === -1) {
-            // The current transform will transform a square pixel into a
-            // parallelogram where both heights are lower than 1 and not equal.
+          if (lineWidth < 0 && -lineWidth >= this.current.lineWidth) {
+            // The current transform will transform a square pixel into
+            // a parallelogram where both heights are lower than 1 and
+            // not equal.
```
Comment (Snuffleupagus) [inline src/display/canvas.js:1378]: Nit: This looks like an unrelated change, please revert it.

File: src/display/canvas.js [hunk 3]
```diff
@@ -2664,25 +2670,31 @@ const CanvasGraphics = (function CanvasGraphicsClosure() {
...
+          // In this case, we must reset the transform, set linewidth to 1
           // and then stroke.
-          this._cachedGetSinglePixelWidth = -1;
-        } else if (sqDet > Number.EPSILON ** 2) {
-          // The multiplication by the constant 1.0000001 is here to have
+          this._cachedGetSinglePixelWidth = -(
+            this._combinedScaleFactor * pixelHeight
+          );
+        } else if (absDet > Number.EPSILON) {
+          // The multiplication by the constant 1.000001 is here to have
```
Comment (Snuffleupagus) [inline src/display/canvas.js:2695]: The constant doesn't actually agree with the code now, please revert this change.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
